Code/CIPLabeler/Node.cpp

```diff
@@ -0,0 +1,142 @@
+//  of the RDKit source tree.
+//
+#include <vector>
+
+#include "Digraph.h"
+#include "Edge.h"
+#include "Node.h"
+#include "CIPMol.h"
+
+namespace RDKit {
+namespace CIPLabeler {
+
+Node *Node::newTerminalChild(int idx, Atom *atom, int flags) const {
+  int new_dist = flags & DUPLICATE ? d_visit[idx] : d_dist + 1;
+  std::vector<char> new_visit;
+
+  if (flags & BOND_DUPLICATE) {
+    auto frac = dp_g->getMol().getFractionalAtomicNum(dp_atom);
+    if (frac.denominator() > 1) {
+      return &dp_g->addNode(std::move(new_visit), atom, std::move(frac),
+                            new_dist, flags);
+    }
+  }
+
+  auto atomic_num = atom ? atom->getAtomicNum() : 1;
+  return &dp_g->addNode(std::move(new_visit), atom, atomic_num, new_dist,
+                        flags);
+}
+
+Node::Node(Digraph *g, std::vector<char> &&visit, Atom *atom,
+           boost::rational<int> &&frac, int dist, int flags)
+    : dp_g{g}, dp_atom{atom}, d_dist{dist}, d_atomic_num{std::move(frac)},
+      d_flags{flags}, d_visit{std::move(visit)} {
+  if (d_flags & DUPLICATE) {
+    d_edges.reserve(4);
+  };
+  if (d_visit.empty() || d_flags & DUPLICATE) {
+    d_flags |= EXPANDED;
+  }
+}
+
+Digraph *Node::getDigraph() const { return dp_g; }
+
+Atom *Node::getAtom() const { return dp_atom; }
+
+int Node::getDistance() const { return d_dist; }
+
+boost::rational<int> Node::getAtomicNumFraction() const { return d_atomic_num; }
+
+int Node::getAtomicNum() const {
+  if (dp_atom == nullptr) {
+    return 1;
+  }
+  return dp_atom->getAtomicNum();
+};
+
+double Node::getMassNum() const {
+  if (dp_atom == nullptr) {
+    return 0;
+  }
+  return dp_atom->getIsotope();
+}
+
+Descriptor Node::getAux() const { return d_aux; }
+
+bool Node::isSet(int mask) const { return mask & d_flags; }
+
+bool Node::isDuplicate() const { return d_flags & DUPLICATE; }
+
+bool Node::isTerminal() const {
+  return d_visit.empty() || (isExpanded() && d_edges.size() == 1);
+}
+
+bool Node::isExpanded() const { return d_flags & EXPANDED; }
+
+bool Node::isVisited(int idx) const { return d_visit[idx] != 0; }
+
+Node *Node::newChild(int idx, Atom *atom) const {
+  auto new_visit = d_visit;
+  new_visit[idx] = static_cast<char>(d_dist + 1);
+  auto atomic_num = atom ? atom->getAtomicNum() : 1;
+  return &dp_g->addNode(std::move(new_visit), atom, atomic_num, d_dist + 1, 0);
+}
+
+Node *Node::newBondDuplicateChild(int idx, Atom *atom) const {
+  return newTerminalChild(idx, atom, BOND_DUPLICATE);
+}
+
+Node *Node::newRingDuplicateChild(int idx, Atom *atom) const {
+  return newTerminalChild(idx, atom, RING_DUPLICATE);
+}
+
+Node *Node::newImplicitHydrogenChild() const {
+  return newTerminalChild(-1, nullptr, IMPL_HYDROGEN);
+}
+
+void Node::add(Edge *e) { d_edges.push_back(e); }
+
+void Node::setAux(Descriptor desc) { d_aux = desc; }
+
+const std::vector<Edge *>& Node::getEdges() const {
+  if (!isExpanded()) {
+    auto non_const_this = const_cast<Node *>(this);
```

Comment:
I also don't love `const_cast`s, and `getEdges()` doesn't have to be `const`, of course, but since expanding the node (calculating the edges) should happen only once for each node, it looks like something reasonable to do.

Reply:
It's your call

Reply:
Ok, I think I will leave it as it is now.